File: helpers/math.ts
import { Position } from "..";

function degreeToRad(degree: number): number {
  return Math.PI * degree / 180;
}

function normalizeDegrees(degree: number): number {
  let angle = degree % 360;
  if (angle < 0) angle += 360;
  return angle;
}

function rotatePoint(point: Position, origin: Position, radAngle: number): Position {
  return {
      x: Math.cos(radAngle) * (point.x-origin.x) - Math.sin(radAngle) * (point.y-origin.y) + origin.x,
      y: Math.sin(radAngle) * (point.x-origin.x) + Math.cos(radAngle) * (point.y-origin.y) + origin.y
  };
}

function checkBoundariesOvelap(a: Array<Position>, b: Array<Position>): boolean {
  for (let i = 0; i < a.length; i++) {
    const lineA: Array<Position> = [
      a[i],
      (i < a.length - 1) ? a[i+1] : a[0]
    ];
    for (let j = 0; j < b.length; j++) {
      const lineB: Array<Position> = [
        b[j],
        (j < b.length - 1) ? b[j+1] : b[0]
      ];

      if (checkLineOverlap(lineA, lineB)) {
        return true;
      }
    }
  }
  return false;
}

function checkLineOverlap(a: Array<Position>, b: Array<Position>): boolean {
  if (a.length !== 2) throw new Error("a is not a valid line")
  if (b.length !== 2) throw new Error("b is not a valid line")
  
  const i: number = (a[1].y - a[0].y)/(a[1].x - a[0].x);
  const j: number = (a[1].x*a[0].y - a[0].x*a[1].y)/(a[1].x - a[0].x);
  const k: number = (b[1].y - b[0].y)/(b[1].x - b[0].x);
  const l: number = (b[1].x*b[0].y - b[0].x*b[1].y)/(b[1].x - b[0].x);
  
  const x: number = (l - j)/(i - k);
  const y: number = (i*l- j*k) /(i - k);

  const overlapPoint: Position = { x, y };

  return isPointInLine(overlapPoint, a[0], a[1]) && isPointInLine(overlapPoint, b[0], b[1]);
}

function isPointInLine(point: Position, lineA: Position, lineB: Position): boolean {
  return getPointDistance(lineA, lineB) == getPointDistance(lineA, point) + getPointDistance(lineB, point);
}

function getPointDistance(a: Position, b: Position): number {
  return Math.sqrt(Math.pow((a.x - b.x), 2) + Math.pow((a.y - b.y), 2));
}

export {
  degreeToRad,
  normalizeDegrees,
  rotatePoint,
  checkBoundariesOvelap,
}

File: rust/helpers/math.rs
use std::f64;
use crate::interfaces::{Angle, Position};

pub fn degree_to_rad(degree: f64) -> f64 {
    std::f64::consts::PI * degree / 180.0
}

pub fn normalize_degrees(degree: f64) -> f64 {
    let mut angle = degree % 360.0;
    if angle < 0.0 {
        angle += 360.0;
    }
    angle
}

pub fn rotate_point(point: Position, origin: Position, angle: Angle) -> Position {
    Position {
        x: angle.cos * (point.x - origin.x) - angle.sin * (point.y - origin.y) + origin.x,
        y: angle.sin * (point.x - origin.x) + angle.cos * (point.y - origin.y) + origin.y,
    }
}

pub fn check_boundaries_overlap(a: &[Position], b: &[Position]) -> bool {
    for i in 0..a.len() {
        let line_a = [
            a[i],
            if i < a.len() - 1 { a[i + 1] } else { a[0] },
        ];
        for j in 0..b.len() {
            let line_b = [
                b[j],
                if j < b.len() - 1 { b[j + 1] } else { b[0] },
            ];

            if check_line_overlap(&line_a, &line_b) {
                return true;
            }
        }
    }
    false
}

pub fn check_line_overlap(a: &[Position], b: &[Position]) -> bool {
    if a.len() != 2 || b.len() != 2 {
        panic!("Invalid line segments");
    }

    let i = (a[1].y - a[0].y) / (a[1].x - a[0].x);
    let j = (a[1].x * a[0].y - a[0].x * a[1].y) / (a[1].x - a[0].x);
    let k = (b[1].y - b[0].y) / (b[1].x - b[0].x);
    let l = (b[1].x * b[0].y - b[0].x * b[1].y) / (b[1].x - b[0].x);

    let x = (l - j) / (i - k);
    let y = (i * l - j * k) / (i - k);

    let overlap_point = Position { x, y };

    is_point_in_line(overlap_point, a[0], a[1]) && is_point_in_line(overlap_point, b[0], b[1])
}

pub fn is_point_in_line(point: Position, line_a: Position, line_b: Position) -> bool {
    (get_point_distance(line_a, line_b) - get_point_distance(line_a, point) - get_point_distance(line_b, point)).abs() < 0.01
}

pub fn get_point_distance(a: Position, b: Position) -> f64 {
    ((a.x - b.x).powi(2) + (a.y - b.y).powi(2)).sqrt()
}
